File: Structer.py
# This Python file uses the following encoding: utf-8

import re


class Structer:
    def __init__(self):
        self.structs = {"STRUCT_DUMMY": "struct DUMMY {}"}
        pass


    def get_structs(self):
        return self.structs


    def readStructs(self, file_path):
        self.structs.clear()
        with open(file_path) as f:
            cur_struct_name = None
            struct_body = False
            for line in f:
                line = line.rstrip("\n")
                if line.startswith("struct ") and line.endswith("{"):
                    regex = re.compile(r'(struct [a-zA-Z0-9-_]+[ ][\{])')
                    mo = regex.search(line)
                    if mo:
                        if len(mo.group(0)) > 0:
                            line = line.strip()
                            line = line.replace("struct s", "struct STRUCT_")
                            struct_name = line.split()[1]
                            self.structs[struct_name] = [line] # probably correct function name, without return type
                            cur_struct_name = struct_name
                            struct_body = True
                elif struct_body:
                    self.structs[cur_struct_name].append(line)
                    if line == "};":
                        struct_body = False

File: CPP_Proj/structer.cpp
#include "structer.h"

void Structer::readStructs(QString filePath)
{
  this->_structs.clear();

  cout << "Reading structs...";

  QRegExp regex1(R"(struct [a-zA-Z0-9-_]+[ ][\{])");

  QFile file(filePath);
  if(!file.open(QIODevice::ReadOnly)) {
      cout << "ERROR: File could not be opened - " << filePath.toStdString().c_str() << endl;
  }

  bool structBody = false;
  QString curSturctName = nullptr;
  QTextStream in(&file);
  while(!in.atEnd()) {
      QString line = in.readLine().remove(QRegExp("([\n]+)$"));
      if (line.startsWith("struct ") && line.endsWith("{")) {
          if((regex1.indexIn(line, 0)) >= 0) {
              //cap = regex1.cap(1).trimmed();
              line = line.trimmed();
              line = line.replace("struct s", "struct STRUCT_");
              curSturctName = line.split(' ')[1];
              this->_structs[curSturctName] = QStringList() << line; // probably correct function name, without return type
              structBody = true;
          }
      } else if (structBody) {
          this->_structs[curSturctName].append(line);
          if (line == QString("};"))
            structBody = false;
      }
  }
  file.close();

  cout << "DONE" << endl;
}
